crates/rolldown/src/module_finalizers/mod.rs

```diff
@@ -328,18 +328,76 @@ impl<'me, 'ast> ScopeHoistingFinalizer<'me, 'ast> {
+        true
+      } else if module_namespace_included_reason
+        .contains(ModuleNamespaceIncludedReason::ReExportExternalModule)
+      {
+        // If the module namespace is only used to reexport external module,
+        // then we need to ensure if it is still has dynamic exports after flatten entry level
+        // external module, see `find_entry_level_external_module`
+        self.ctx.linking_info.has_dynamic_exports
+      } else {
+        false
+      };
+
+    if !is_namespace_referenced {
+      return vec![];
+    }
+
     let binding_name_for_namespace_object_ref =
       self.canonical_name_for(self.ctx.module.namespace_object_ref);
-    // construct `var [binding_name_for_namespace_object_ref] = {}`
-    let decl_stmt = self.snippet.var_decl_stmt(
-      binding_name_for_namespace_object_ref,
-      ast::Expression::ObjectExpression(ArenaBox::new_in(
-        ObjectExpression::dummy(self.alloc),
-        self.alloc,
-      )),
-    );
 
-    let has_exports = self.ctx.linking_info.canonical_exports(false).next().is_some();
+    // construct `{ prop_name: () => returned, ... }`
+    let mut arg_obj_expr = ast::ObjectExpression::dummy(self.alloc);
+
+    arg_obj_expr.properties.extend(self.ctx.linking_info.canonical_exports(false).map(
+      |(export, resolved_export)| {
+        // prop_name: () => returned
+        let prop_name = export;
+        let returned = self.finalized_expr_for_symbol_ref(resolved_export.symbol_ref, false, false);
+        ast::ObjectPropertyKind::ObjectProperty(
+          ast::ObjectProperty {
+            key: if is_validate_identifier_name(prop_name) {
+              ast::PropertyKey::StaticIdentifier(
+                self.snippet.id_name(prop_name, SPAN).into_in(self.alloc),
+              )
+            } else {
+              ast::PropertyKey::StringLiteral(self.snippet.alloc_string_literal(prop_name, SPAN))
+            },
+            value: self.snippet.only_return_arrow_expr(returned),
+            ..ast::ObjectProperty::dummy(self.alloc)
+          }
+          .into_in(self.alloc),
+        )
+      },
+    ));
+
+    // if there is no export, we should generate `var ns = {}` instead of `var ns = __export({}, {})`
+    // else construct `__export(ns_name, { prop_name: () => returned, ... })`
+    let module_namespace_rhs = if arg_obj_expr.properties.is_empty() {
+      Expression::ObjectExpression(self.builder().alloc(arg_obj_expr))
+    } else {
+      self.snippet.builder.expression_call(
+        SPAN,
+        self.finalized_expr_for_runtime_symbol("__export"),
+        NONE,
+        self.snippet.builder.vec_from_array([
+          ast::Argument::from(ast::Expression::ObjectExpression(ArenaBox::new_in(
+            ObjectExpression::dummy(self.alloc),
+            self.alloc,
+          ))),
+          ast::Argument::ObjectExpression(arg_obj_expr.into_in(self.alloc)),
+        ]),
+        false,
+      )
```

Comment:
```suggestion
      self.snippet.builder.expression_call_with_pure(
        SPAN,
        self.finalized_expr_for_runtime_symbol("__export"),
        NONE,
        self.snippet.builder.vec_from_array([
          ast::Argument::from(ast::Expression::ObjectExpression(ArenaBox::new_in(
            ObjectExpression::dummy(self.alloc),
            self.alloc,
          ))),
          ast::Argument::ObjectExpression(arg_obj_expr.into_in(self.alloc)),
        ]),
        false,
        true,
      )
```
I think we want to mark this as pure.

Reply:
Ah, maybe in a separate PR.

Reply:
Yeah, you are right, I am going to do this with another pr.